### ISSUE TITLE
Getting this into postmarketOS

### ISSUE THREAD

**Issue by kcxt** [michele-perrone/postmarketos-oneplus6-tristatekey#1]
Heya! I'm super excited at how fast this has been developed, I'm really looking forward to using it.

It looks like it's a bit more complicated to deal with than I first expected, a few potential issues:

- Adding the user to the gpiod group gives them permissiong to access all of the GPIOs, this is probably not great for security.
- Installing the .desktop file to the user home directory isn't doable, you use `$USER` in the APKBUILD which doesn't work.
    - Instead, you could install to /usr/share/applications and use a post-install script to put it somewhere like `etc/xdg/autostart/` see https://help.gnome.org/admin/system-admin-guide/stable/autostart-applications.html.en

The issue with dbus and users is a hard one, perhaps just having the daemon run as root and use uinput to generate input events which feedbackd can then handle? Alternatively a secondary user daemon could talk to the root daemon and propagate stuff to feedbackd with dbus. bleh

Missings deps in APKBUILD: `libgpiod-dev` and `glib-dev` build time dep.

**Comment by michele-perrone** [comment]
Hi Caleb! Thank you for your feedback and encouragement.
- Thanks for pointing out the mistakes in `APKBUILD` about the `.desktop` file and build time dependencies. It's my first Alpine package so I'm getting up to speed.  I'll make sure to fix them.
- Yes, I agree, giving permission to read/write all of the GPIOs is not the best thing to do. I guess I could have given permissions for `gpiochip2` only , but as far as I know, with the `.rules` file we can't give permission for single GPIO lines - only for entire GPIO chips. So this is a good argument for running the daemon as root.
- I like the idea of the secondary user daemon started by the main root daemon. I'll dig into that. If I succeed, then we could go also back to OpenRC.

**Comment by michele-perrone** [comment]
@calebccff can I mark this as closed? Please let me know.

**Comment by kcxt** [comment]
Hey! I guess pretty much everything is done. Are you ready to open an MR to https://gitlab.com/postmarketOS/pmaports? You will need to adjust the APKBUILD so that it can be included in pmaports and will fetch this git repo and do the build, here's an example: https://gitlab.com/postmarketOS/pmaports/-/blob/master/main/fbkeyboard/APKBUILD

Imho there is no reason to have an APKBUIILD in this repository, as it will not match the APKBUILD which would actually be used to build the package for postmarketOS.

**Comment by michele-perrone** [comment]
Ok, I'm back and I'm really sorry that it's taking so long despite my initial sprint. I just couldn't find the time.   
1. Yes of course, I'll adjust the APKBUILD the way you said, remove it from this repository and submit the MR.
2. Once that's done, I'll start doing the following: 
2.1 modifying the root daemon to read the GPIO chip and lines from a configuration file. This will be done only once
2.2 modifying the user daemon to run arbitrary commands based on the position of the tri-state key. Configuration file can be modified while the user daemon is running
2.3 integration with the Tweaks app

I'll take care of point 1. by the end of the week, tops. I promise.

**Comment by kcxt** [comment]
hey, that's awesome! dont worry theres no rush to do anything, i really appreciate this work!

i did some research into input devices and found out that theres a whole inout event type for "misc" devices which is perfect for us, so it should be possible to quite easily move the root daemon parts into the kernel and expose the tri-state-key as an input device with udev rules to give the user permissions to access it.

let me know what you think, that should simplify things a lot afaict.

**Comment by michele-perrone** [comment]
Yes, the kernel driver would be a good step towards distribution-independent code, because we wouldn't need to worry about the presence `systemd`, `openRC`, or other `init` systems.

If I understood you correctly, this would involve writing a Misc Device Driver (major code 10) that reads the `GPIO` pins. The user daemon then simply opens the device and reads the tri-state key value.

There's just one thing that worries me. Right now, we're using a blocking pipe to handle the root-user daemon communication. This is actually cool, because it removes the need for the user daemon to poll the current state. Now with the Misc Device Driver, is there a way to trigger an interrupt, so the user daemon doesn't call `read()` all the time? (Forgive my ignorance)

**Comment by kcxt** [comment]
it should just work with the gpio-keys driver in the kernel. you can setup call backs for input events so i don't think you'd need to do any polling.

**Comment by phodina** [comment]
Hi, wanted to access the keys directly and look into using the interrupts.

However, seems the GPIOs are handled by some other process/kernel.

```
pocket:~$ sudo gpioget -c gpiochip2 24
gpioget: unable to request lines: Resource busy
pocket:~$ sudo gpioget -c gpiochip2 52
gpioget: unable to request lines: Resource busy
pocket:~$ sudo gpioget -c gpiochip2 126
gpioget: unable to request lines: Resource busy
cat /proc/device-tree/model 
OnePlus 6pocket
```

And seems it's integrated already in the system:
```
cat /proc/bus/input/devices
I: Bus=0000 Vendor=0000 Product=0000 Version=0000
N: Name="pm8941_pwrkey"
P: Phys=pm8941_pwrkey/input0
S: Sysfs=/devices/platform/soc@0/c440000.spmi/spmi-0/0-00/c440000.spmi:pmic@0:pon@800/c440000.spmi:pmic@0:pon@800:pwrkey/input/input0
U: Uniq=
H: Handlers=kbd event0 
B: PROP=0
B: EV=3
B: KEY=10000000000000 0

I: Bus=0019 Vendor=0001 Product=0001 Version=0100
N: Name="Alert slider"
P: Phys=gpio-keys/input0
S: Sysfs=/devices/platform/alert-slider/input/input1
U: Uniq=
H: Handlers=event1 
B: PROP=0
B: EV=9
B: ABS=400000000

pocket:~$ sudo evtest /dev/input/event1
Input driver version is 1.0.1
Input device ID: bus 0x19 vendor 0x1 product 0x1 version 0x100
Input device name: "Alert slider"
Supported events:
  Event type 0 (EV_SYN)
  Event type 3 (EV_ABS)
    Event code 34 (?)
      Value      1
      Min        0
      Max        0
Properties:
Testing ... (interrupt to exit)
Event: time 1744529855.956810, type 3 (EV_ABS), code 34 (?), value 2
Event: time 1744529855.956810, -------------- SYN_REPORT ------------
Event: time 1744529859.866623, type 3 (EV_ABS), code 34 (?), value 1
Event: time 1744529859.866623, -------------- SYN_REPORT ------------
Event: time 1744529860.894986, type 3 (EV_ABS), code 34 (?), value 0
Event: time 1744529860.894986, -------------- SYN_REPORT ------------
Event: time 1744529862.800895, type 3 (EV_ABS), code 34 (?), value 1
Event: time 1744529862.800895, -------------- SYN_REPORT ------------
Event: time 1744529863.475299, type 3 (EV_ABS), code 34 (?), value 2
Event: time 1744529863.475299, -------------- SYN_REPORT ------------
```
